packages/pwa-kit-react-sdk/src/ssr/universal/components/route-component/index.js

```diff
@@ -402,17 +402,26 @@ export const routeComponent = (Wrapped, isPage, locals) => {
  *
  * @private
  */
-export const getRoutes = (locals = {}) => {
+export const getRoutes = async (locals = {}, req = {}) => {
     let _routes = routes
     const {applicationExtensions = []} = locals
     if (typeof routes === 'function') {
-        _routes = routes()
+        _routes = await routes(locals)
     }
 
     // Call the `extendRoutes` function for all the Application Extensions.
-    applicationExtensions.forEach((applicationExtension) => {
-        _routes = applicationExtension.extendRoutes(_routes)
-    })
+    for (const applicationExtension of applicationExtensions) {
+        const routes = await applicationExtension.extendRoutes(_routes, req)
+        const extensionName = applicationExtension.constructor.name
+        
+        // Prefix each component displayName with the extension name so it can later be deserialized
+        routes.forEach((route) => {
+            // Skip if component is already prefixed with an application extension name
+            if (route.component.displayName.includes(".") && route.component.displayName.match(/^[^.]+/)[0]) return
+            route.component.displayName = `${extensionName}.${route.component.displayName}`
+        })
```

Comment:
because `extendRoutes` modifies the original `routes` array, here I prefix the component displayName only when it  doesn't have a prefix already. So any new routes added via an applicationExtension `extendRoutes` call will get the extension prefix added